### PR TITLE
Add MTP acceptance rate metrics

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?
Refactor `MTPLossLoggingHelper` to track acceptance rate (correct/total predictions) alongside loss metrics. Computes distributed argmax across tensor parallel ranks to determine predictions without gathering full vocab logits (conditionally). Logs per-step and cumulative acceptance rates.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/multi_token_prediction.py [hunk 1]
```diff
@@ -337,80 +337,213 @@ def _roll_tensor_packed_seq(tensor, shifts, dims, packed_seq_params, cp_group=No
...
+    mtp_logits: Tensor,
+    mtp_labels: Tensor,
+    loss_mask: Tensor,
+    output_layer: Callable,
+    runtime_gather_output: Optional[bool],
+) -> Tuple[Tensor, Tensor]:
+    """Compute MTP acceptance correct/total counts."""
+    with torch.no_grad():
+        tp_group = parallel_state.get_tensor_model_parallel_group()
+        tp_size = parallel_state.get_tensor_model_parallel_world_size()
```
Comment (santhnm2) [inline megatron/core/transformer/multi_token_prediction.py]: Should we be passing a `tp_group` here instead of using `parallel_state`?

File: megatron/core/transformer/multi_token_prediction.py [hunk 2]
```diff
@@ -337,80 +337,213 @@ def _roll_tensor_packed_seq(tensor, shifts, dims, packed_seq_params, cp_group=No
 
 
 class MTPLossLoggingHelper:
-    """Helper class for logging MTP losses."""
+    """Helper class for logging MTP losses and acceptance rates."""
```
Comment (santhnm2) [inline megatron/core/transformer/multi_token_prediction.py:343]: This comment is not necessary to address for this PR itself, but eventually we will want to unify the MTP acceptance rate tracking for training and inference - for that effort, maybe there is some value in keeping the loss tracking separate from the acceptance rate tracking so that the latter can be shared while the former is used only for training.

File: megatron/core/transformer/multi_token_prediction.py [hunk 3]
```diff
@@ -337,80 +337,213 @@ def _roll_tensor_packed_seq(tensor, shifts, dims, packed_seq_params, cp_group=No
...
-            loss = mtp_losses[i]
+            loss_name = f"mtp_{i+1} loss"
+            step_acc_name = f"mtp_{i+1}_acceptance_rate"
+            cum_acc_name = f"mtp_{i+1}_cumulative_acceptance_rate"
+
+            loss = mtp_loss_sums[i] / torch.clamp(mtp_num_tokens[i], min=1)
+            step_rate = (mtp_corrects[i] / torch.clamp(mtp_totals[i], min=1)) * 100.0
+            cum_rate = (
+                mtp_cumulative_corrects[i] / torch.clamp(mtp_cumulative_totals[i], min=1)
+            ) * 100.0
```
Comment (santhnm2) [inline megatron/core/transformer/multi_token_prediction.py]: Are these clamps necessary to prevent divide by zero errors? If so could you add a comment to clarify?

**Review by santhnm2 — COMMENTED** [review]
<ignore>

File: megatron/core/transformer/multi_token_prediction.py (hunk 2 above)
Comment (macandro96) [reply megatron/core/transformer/multi_token_prediction.py:343]: Makes sense. We have this function [here](https://github.com/yfw/Megatron-LM/blob/d6d1eeae1881f76c8e900e5dd09d121217dfc9fc/megatron/core/transformer/multi_token_prediction.py#L517) to calculate acceptance metrics that can be reused when things separate out.

File: megatron/core/transformer/multi_token_prediction.py (hunk 1 above)
Comment (macandro96) [reply megatron/core/transformer/multi_token_prediction.py]: Added!

File: megatron/core/transformer/multi_token_prediction.py (hunk 3 above)
Comment (macandro96) [reply megatron/core/transformer/multi_token_prediction.py]: done!

**Review by yashaswikarnati — APPROVED** [review]

**Review by santhnm2 — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
